whisper_medusa/dataset/dataset.py

```diff
@@ -0,0 +1,171 @@
+import logging
+from dataclasses import dataclass
+from typing import Dict, List, Union
+import pandas as pd
+import torch
+import torchaudio
+from transformers import WhisperProcessor
+from whisper_medusa.utils.utils import token_padding
+
+SAMPLE_RATE = 16_000
+
+
+class ASRDataSet(torch.utils.data.Dataset):
+    """
+    ASR Dataset
+
+    Parameters
+    ----------
+    data_path : str. Path to csv file with columns: "sentence", "path".
+    """
+
+    def __init__(
+        self,
+        data_path,
+        split,
+        processor: WhisperProcessor,
+        target_sample=16_000,
+        debug_mode=False,
+    ):
+        assert split in [
+            "train",
+            "val",
+            "test",
+        ]  # sanity for case we will use split later.
+        self.split = split
+        self.debug_mode = debug_mode
+
+        # read csv file + create dataset
+        self.data_path = data_path
+        self.dataset_df = pd.DataFrame()
+        self._init_dataset_obj()
+
+        self.dataset = self.dataset_df.to_dict("records")
+        self.length = len(self.dataset)
+
+        self.target_sample_rate = target_sample
+
+        self.processor = processor
+
+    def _init_dataset_obj(self):
+        self.dataset_df = pd.read_csv(self.data_path)
+        self.dataset_df.sentence = self.dataset_df.sentence.fillna("")
+
+
+        # in case of debug mode, sample only 1000 samples
+        if self.debug_mode and self.split in ["val", "test"]:
+            sample_size = min(1000, len(self.dataset_df))
+            self.dataset_df = self.dataset_df.sample(n=sample_size)
+        self.dataset = self.dataset_df.to_dict("records")
+
+        # check case language is not specified
+        if "language" not in self.dataset_df.columns:
+            logging.info("[NOTE]: No language specified, using tokenizer's language")
+
+    def __len__(self):
+        return self.length
+
+    @staticmethod
+    def speech_file_to_array(path, resampling_to=16_000):
+        batch = {}
+        # todo: consider using to soundfile to prevent issues with mp3 files with 'ffmpeg>=5'.
+        #  If you encounter issues with mp3 files use `conda install 'ffmpeg<5'`
+        speech_array, sampling_rate = torchaudio.load(path)
+        resampler = torchaudio.transforms.Resample(sampling_rate, resampling_to)
+        array = resampler(speech_array)[0].numpy()
+        batch["audio"] = {"array": array, "sampling_rate": resampling_to}
+        return batch
+
+    def prepare_dataset(self, batch):
+        audio = batch["audio"]
+
+        # compute log-Mel input features from input audio array
+        # todo: switch to our feature_extractor implementation which is faster.
```

Comment:
remove todo, and consider switching to our implementation.

Reply:
done